zkevm-circuits/src/evm_circuit/util/precompile_gadget.rs

```diff
@@ -0,0 +1,99 @@
+use bus_mapping::precompile::PrecompileCalls;
+use eth_types::Field;
+use gadgets::util::Expr;
+use halo2_proofs::plonk::Expression;
+
+use crate::evm_circuit::step::ExecutionState;
+
+use super::{
+    constraint_builder::{ConstrainBuilderCommon, EVMConstraintBuilder},
+    math_gadget::BinaryNumberGadget,
+    CachedRegion,
+};
+
+#[derive(Clone, Debug)]
+pub struct PrecompileGadget<F> {
+    address: BinaryNumberGadget<F, 4>,
+}
+
+impl<F: Field> PrecompileGadget<F> {
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn construct(
+        cb: &mut EVMConstraintBuilder<F>,
+        is_success: Expression<F>,
+        callee_address: Expression<F>,
+        _caller_id: Expression<F>,
+        _cd_offset: Expression<F>,
+        cd_length: Expression<F>,
+        _rd_offset: Expression<F>,
+        _rd_length: Expression<F>,
+        precompile_return_length: Expression<F>,
+        // input bytes to precompile call.
+        input_bytes_rlc: Expression<F>,
+        // output result from precompile call.
+        output_bytes_rlc: Expression<F>,
+        // returned bytes back to caller.
+        _return_bytes_rlc: Expression<F>,
```

Comment:
do you think we will need `return_bytes_rlc/rd_length` later?   If PrecompileGadget is designed to handle precompile contract computation, then slice output_bytes to return_bytes by rd_length is the duty of the caller?   So PrecompileGadget does not need to know about `return_bytes_rlc/rd_length`?